### PR TITLE
Move dictionary declarations of Alpaka serial backend data products to the host classes_def.xml

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR moves the dictionary declarations of Alpaka serial backend data products to the host `classes_def.xml`. The "serial" data products are intended to be consumable by non-Alpaka EDModules as well, so having all of their definitions in the main library is more consistent. See more discussion in https://github.com/cms-sw/cmssw/pull/40690#discussion_r1096102877 .

I noticed though that without the `classes_serial_def.xml` and `classes_serial.h` files the build fails with
```
>> Building LCG reflex dict from header file src/DataFormats/PortableTestObjects/src/alpaka/classes_serial.h
Error: Cannot find header /build/mkortela/coresw/CMSSW_13_0_0_pre4//src/DataFormats/PortableTestObjects/src/alpaka/classes_serial.h: cannot inline it.
Error: rootcling: cannot open linkdef file src/DataFormats/PortableTestObjects/src/alpaka/classes_serial_def.xml
gmake: *** [config/SCRAM/GMake/Makefile.rules:1740: tmp/slc7_amd64_gcc11/src/DataFormats/PortableTestObjects/src/alpaka/DataFormatsPortableTestObjectsSerialSync/a/DataFormatsPortableTestObjectsSerialSync_xr.cc] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```
and just to get started I included both with dummy content.

#### PR validation:

Unit tests pass (on a machine without GPU)

### CODE REVIEW

**Comment by makortel** [comment]
hold

(maybe I should have opened this as draft...)

**Comment by makortel** [comment]
enable gpu

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@smuzaffar Would it be feasible to modify the Alpaka library build rules such that the dictionaries for the `serial` backend are generated only if the `classes_serial.h` or `classes_serial_def.xml` exist?

Alternatively we could decide that we'd never generate dictionaries for the `serial` backend, but I'd be tempted to keep the option in case some unexpected use case comes up.

**Comment by makortel** [comment]
(@fwyzard I'll rebase this PR on top of the commit of #40789 later)

**Comment by makortel** [comment]
So what actually happens in my test when I remove the `classes_serial_def.xml` and `classes_serial.h` and do the `scram b clean`, the building of `DataFormats/PortableTestObjects`  itself works fine.

But the build gets later stuck in `edmWriteConfigs -p /build/mkortela/coresw/CMSSW_13_0_0_pre4/tmp/slc7_amd64_gcc11/src/HeterogeneousCore/AlpakaTest/plugins/HeterogeneousCoreAlpakaTestPluginsPortableSerialSync/libHeterogeneousCoreAlpakaTestPluginsPortableSerialSync.so`. The `edmWriteConfigs` itself is stuck in
```
#0  0x00007fc7df13154d in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007fc7df12ce9b in _L_lock_883 () from /lib64/libpthread.so.0
#2  0x00007fc7df12cd68 in pthread_mutex_lock () from /lib64/libpthread.so.0
#3  0x00007fc7e01dd51e in DebugPrint(char const*, ...) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#4  0x00007fc7e01dd7ed in DefaultErrorHandler(int, bool, char const*, char const*) ()
   from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#5  0x00007fc7e0296c25 in ErrorHandler () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#6  0x00007fc7e01f013c in TObject::Warning(char const*, char const*, ...) const ()
   from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#7  0x00007fc7d8562da1 in TCling::ReadRootmapFile(char const*, TCling::TUniqueString*) ()
   from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCling.so
#8  0x00007fc7d8597d23 in TCling::LoadLibraryMap(char const*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCling.so
#9  0x00007fc7d8598bf7 in TCling::Initialize() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCling.so
#10 0x00007fc7e01b1fed in TROOT::InitInterpreter() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#11 0x00007fc7e01b233f in ROOT::Internal::GetROOT2() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#12 0x00007fc7e01dc04e in TEnv::Getvalue(char const*) const () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#13 0x00007fc7e01dc949 in TEnv::GetValue(char const*, char const*) const ()
   from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#14 0x00007fc7e01ddb18 in DefaultErrorHandler(int, bool, char const*, char const*) ()
   from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#15 0x00007fc7e0296c25 in ErrorHandler () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#16 0x00007fc7e02975a8 in Warning(char const*, char const*, ...) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#17 0x00007fc7e02c851f in ROOT::TGenericClassInfo::TGenericClassInfo(char const*, int, char const*, int, std::type_info const&, ROOT::Internal::TInitBehavior const*, TClass* (*)(), TVirtualIsAProxy*, int, int) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/external/slc7_amd64_gcc11/lib/libCore.so
#18 0x00007fc7dc052989 in ROOT::GenerateInitInstanceLocal(edm::Wrapper<PortableHostCollection<portabletest::TestSoALayout<128ul, false> > > const*) [clone .constprop.0] ()
   from /build/mkortela/coresw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libDataFormatsPortableTestObjects.so
#19 0x00007fc7dc05240c in _GLOBAL__sub_I_DataFormatsPortableTestObjects_xr.cc ()
   from /build/mkortela/coresw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libDataFormatsPortableTestObjects.so
#20 0x00007fc7e0eab9c3 in _dl_init_internal () from /lib64/ld-linux-x86-64.so.2
#21 0x00007fc7e0eb059e in dl_open_worker () from /lib64/ld-linux-x86-64.so.2
#22 0x00007fc7e0eab7d4 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#23 0x00007fc7e0eafb8b in _dl_open () from /lib64/ld-linux-x86-64.so.2
#24 0x00007fc7dfa8afab in dlopen_doit () from /lib64/libdl.so.2
#25 0x00007fc7e0eab7d4 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#26 0x00007fc7dfa8b5ad in _dlerror_run () from /lib64/libdl.so.2
#27 0x00007fc7dfa8b041 in dlopen@@GLIBC_2.2.5 () from /lib64/libdl.so.2
#28 0x00007fc7e0ee2af6 in edmplugin::SharedLibrary::SharedLibrary(std::filesystem::__cxx11::path const&) ()
   from /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libFWCorePluginManager.so
```
which looks the same as in https://github.com/root-project/root/issues/11383.

`ldd /build/mkortela/coresw/CMSSW_13_0_0_pre4/tmp/slc7_amd64_gcc11/src/HeterogeneousCore/AlpakaTest/plugins/HeterogeneousCoreAlpakaTestPluginsPortableSerialSync/libHeterogeneousCoreAlpakaTestPluginsPortableSerialSync.so` shows
```
        libHeterogeneousCoreAlpakaTest.so => /build/mkortela/coresw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libHeterogeneousCoreAlpakaTest.so (0x00007fdd39425000)
        libHeterogeneousCoreAlpakaTestSerialSync.so => /build/mkortela/coresw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libHeterogeneousCoreAlpakaTestSerialSync.so (0x00007fdd39420000)
        libHeterogeneousCoreAlpakaCore.so => /build/mkortela/coresw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libHeterogeneousCoreAlpakaCore.so (0x00007fdd39419000)
        libHeterogeneousCoreAlpakaCoreSerialSync.so => /build/mkortela/coresw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libHeterogeneousCoreAlpakaCoreSerialSync.so (0x00007fdd3940a000)
        libDataFormatsPortableTestObjects.so => /build/mkortela/coresw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libDataFormatsPortableTestObjects.so (0x00007fdd393fd000)
        libDataFormatsPortableTestObjectsSerialSync.so => /cvmfs/cms.cern.ch/slc7_amd64_gcc11/cms/cmssw/CMSSW_13_0_0_pre4/lib/slc7_amd64_gcc11/libDataFormatsPortableTestObjectsSerialSync.so (0x00007fdd393f1000)
```
I guess the dependence on `libDataFormatsPortableTestObjectsSerialSync.so` is a result of `DataFormats/PortableTestObjects/BuildFile.xml` containing `<flags ALPAKA_BACKENDS="1"/>`, i.e. the package would lead to `libDataFormatsPortableTestObjects<backend>.so` shared objects that a dependent packace's library would be linked against (as happens here). And now I'm trying to make the `SerialSync` case special (by removing the library for data formats).

**Comment by makortel** [comment]
Thanks @smuzaffar for the reminder, 
> ```
> <flags ALPAKA_BACKENDS="cuda rocm"/>
> ```

indeed works. For longer term, I'm wondering would it be feasible to add a possibility for an inverted logic, i.e. "all backends except the listed ones"? Although we should re-think the relation of "platform" concept in DataFormats (and ESProducers(?) and Services) and "backend" concept in EDModules, and the impact of that for the library organization and build rules anyway, so I'm not if this detail is worth of addressing before the outcome of the platform-backend thought process.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
(this PR is also now based on top of #40789)

File: DataFormats/PortableTestObjects/BuildFile.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
 <use name="DataFormats/Portable"/>
 <use name="DataFormats/SoATemplate"/>
 <use name="HeterogeneousCore/AlpakaInterface"/>
-<flags ALPAKA_BACKENDS="1"/>
+<flags ALPAKA_BACKENDS="cuda rocm"/>
```
Comment (fwyzard) [inline DataFormats/PortableTestObjects/BuildFile.xml:7]: Assuming this approach of having the "serial" data formats in the main (non-alpaka) directory becomes the preferred one, we should find a way to automate this.

**Comment by makortel** [comment]
@cmsbuild, please test for el8_aarch64_gcc11

File: DataFormats/PortableTestObjects/BuildFile.xml (hunk 1 above)
Comment (makortel) [reply DataFormats/PortableTestObjects/BuildFile.xml:7]: I see @smuzaffar opened an issue https://github.com/cms-sw/cmssw-config/issues/93 to improve the syntax, so how about we proceed now with this approach and invert it later on? (I'd expect the improvement to be available before we would be adding a new backend)

Comment (fwyzard) [reply DataFormats/PortableTestObjects/BuildFile.xml:7]: The next "backend" we could add (even if more for testing purpose than for use in production) would probably be the CPU block-parallel one based on TBB.
If we add that, `<flags ALPAKA_BACKENDS="cuda rocm"/>` would still be the correct choice for any `DataFormats`.

So, yes, I agree we can move ahead with this approach, and improve things later.

**Comment by makortel** [comment]
Rebased on top of #40813

**Comment by makortel** [comment]
unhold

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
+heterogeneous

**Comment by perrotta** [comment]
+1

**Comment by dan131riley** [comment]
We're getting warnings (including in the test jobs for this PR):
```
Warning in <TInterpreter::ReadRootmapFile>: class  PortableHostCollection<portabletest::TestSoALayout<128,false> > found in libDataFormatsPortableTestObjectsSerialSync.so  is already in libDataFormatsPortableTestObjects.so 
Warning in <TInterpreter::ReadRootmapFile>: class  PortableHostCollection<portabletest::TestSoALayout<128ul,false> > found in libDataFormatsPortableTestObjectsSerialSync.so  is already in libDataFormatsPortableTestObjects.so 
Warning in <TInterpreter::ReadRootmapFile>: class  edm::Wrapper<PortableHostCollection<portabletest::TestSoALayout<128,false> > > found in libDataFormatsPortableTestObjectsSerialSync.so  is already in libDataFormatsPortableTestObjects.so 
Warning in <TInterpreter::ReadRootmapFile>: class  edm::Wrapper<PortableHostCollection<portabletest::TestSoALayout<128ul,false> > > found in libDataFormatsPortableTestObjectsSerialSync.so  is already in libDataFormatsPortableTestObjects.so 
Warning in <TInterpreter::ReadRootmapFile>: class  edm::Wrapper<portabletest::TestHostCollection> found in libDataFormatsPortableTestObjectsSerialSync.so  is already in libDataFormatsPortableTestObjects.so 
Warning in <TInterpreter::ReadRootmapFile>: class  portabletest::TestHostCollection found in libDataFormatsPortableTestObjectsSerialSync.so  is already in libDataFormatsPortableTestObjects.so 
```
and unit test failures:
```
Test name: testStandalone::writeAndReadFile
uncaught exception of type std::exception (or derived).
- An exception of category 'FatalRootError' occurred while
   [0] Constructing service of type InitRootHandlers
   Additional Info:
      [a] Fatal Root Error: @SUB=TInterpreter::ReadRootmapFile
class  PortableHostCollection<portabletest::TestSoALayout<128,false> > found in libDataFormatsPortableTestObjectsSerialSync.so  is already in libDataFormatsPortableTestObjects.so 
```
that appear to be related to this PR?

**Comment by makortel** [comment]
I'd expect those to disappear in a full build (the latest IB is still a patch build).

@smuzaffar This PR removed the `libDataFormatsPortableTestObjectsSerialSync.so` (and only that of all the `libDataFormatsPortableTestObjects*.so`). Just guessing, but could it be that this corner case escapes the "poisoned library" setup?

**Comment by smuzaffar** [comment]
> I'd expect those to disappear in a full build (the latest IB is still a patch build).

right, a full build should fix this

> @smuzaffar This PR removed the `libDataFormatsPortableTestObjectsSerialSync.so` (and only that of all the `libDataFormatsPortableTestObjects*.so`). Just guessing, but could it be that this corner case escapes the "poisoned library" setup?

although we have protection against edm plugins (via poison plugin cache) but we do not poison shared libraires itself. I think we can fix this as well by creating poison libs but I am not sure if root will behave (it still might go through the LD_LIBRARY_PATH and load the shared library from release area).

**Comment by makortel** [comment]
Ah, the poison libraries were only for plugins. I started to wonder, because when we have moved dictionaries around in the past, I didn't recall any unit tests to fail because of that (but only the IB duplicate dictionary checker complaining until the next full build).

**Pull request merged**